torch/onnx/_type_utils.py

```diff
@@ -108,36 +114,94 @@ def from_name(
     @classmethod
     @_beartype.beartype
     def from_dtype(cls, dtype: torch.dtype) -> JitScalarType:
-        """Convert a torch dtype to ScalarType."""
+        """Convert a torch dtype to JitScalarType.
+
+        Note: DO NOT USE this API when `dtype` comes from a `torch._C.Value.type()` calls.
+            A "RuntimeError: INTERNAL ASSERT FAILED at "../aten/src/ATen/core/jit_type_base.h" can
+            be raised in several scenarios where shape info is not present.
+            Instead use `from_value` API which is safer.
+
+        Args:
+            dtype: A torch.dtype to create a JitScalarType from
+
+        Returns:
+            JitScalarType
+
+        Raises:
+            ValueError: if dtype is not a valid torch.dtype or if it is None.
+        """
         if dtype not in _DTYPE_TO_SCALAR_TYPE:
             raise ValueError(f"Unknown dtype: {dtype}")
         return _DTYPE_TO_SCALAR_TYPE[dtype]
 
+    # TODO: `value: torch._C.Value` type annotation raises (CI machines only):
+    #       NameError: name '_C' is not defined
+    # In both cases, using typing.TypeVar or string literals didn't work
+    @classmethod
+    @_beartype.beartype
+    def from_value(cls, value, default=None) -> JitScalarType:
```

Comment:
I am going to test the typing and see what happens

Reply:
Yup, there is some dark magic happening here. I've tried `from torch import _C`, `torch._C.Value` and some variations. CI screams on this particular `type_utils.py` for some reason. When the PR is passing all tests and ready for merge, I will try a couple things specifically for this type annotation